### PR TITLE
Support for soft clustering

### PR DESCRIPTION
**Pull request by Moonire**
add's the option to keep overlapping nodes in the clustering or to delete their duplicates by setting keep_overlap either to True or False. Displaying a message informing the user that there are overlaps in case he is using the default hard clustering method.

### CODE REVIEW

File: markov_clustering/mcl.py [hunk 1]
```diff
@@ -137,11 +137,12 @@ def iterate(matrix, expansion, inflation):
     return matrix
 
 
-def get_clusters(matrix):
+def get_clusters(matrix, keep_overlap=False):
     """
     Retrieve the clusters from the matrix
     
     :param matrix: The matrix produced by the MCL algorithm
+    :param keep_overlap: If true, unables soft clustering
```
Comment (GuyAllard) [inline markov_clustering/mcl.py]: use 'enable' instead of 'unable'

File: markov_clustering/mcl.py [hunk 2]
```diff
@@ -161,7 +162,34 @@ def get_clusters(matrix):
...
+    # converting it to a list
+    clusters = sorted(list(clusters))
+
+    # checks for overlaping
+    clusters_total_size = sum(len(c) for c in clusters)
+
+    if matrix.shape[0] < clusters_total_size and keep_overlap == False:
+
+        printer = MessagePrinter(True)
+        printer.print("Clustering contains overlapping, to unable soft clustering set keep_overlap to True")
```
Comment (GuyAllard) [inline markov_clustering/mcl.py]: use 'enable' instead of 'unable'

**Review by GuyAllard — CHANGES_REQUESTED** [review]

**Comment by GuyAllard** [comment]
Hi Mounir, Thanks for submitting this PR.

I would like to approach this slightly differently to your implementation.

Specifically, I would like to perform the identification and removal of overlap nodes in a separate function which is called when keep_overlap=False
Doing so will make the code easier to maintain and test.

What are your thoughts on this approach?

**Comment by Moonire** [comment]
It seems perfectly reasonable to me.
In this case I think a ternary operator that will call the function when keep_overlap=False will do the trick. I'll take care of that asap.

Do you want to change the warning message ? I know i'm not good with writing warnings haha.

**Comment by codykingham** [comment]
It seems a fix for this was never implemented?

**Pull request closed without merging**